### ISSUE TITLE
Nitrokey App should be more accessible to screen reader users

### ISSUE THREAD

**Issue by heikofolkerts** [Nitrokey/nitrokey-app#9]
I am using a screen reader such as NVDA http://www.nvda-project.org to access the contents of applications. The Nitrokey App is accessible to screenreaders to some extend, but there are some small things missing:
- All controls in the dialogs should have the property "accessibleName" set to inform the screen reader user what the control will contain or do. In general this is the label attached to a edit field or list box etc. Since the code seems to be in QT5 setting the property is just a line of code for every control.
- The order in which the controls get focus when using the tab key is restricted to just editable controls. It would be very helpful if the order will contain all (even read only controls) so that the tab key will visit every part of the dialog. When setting up one time passwords reading read only field contents is crutial. If including all controls in this tab order will disturbe non screen reader users then detection of an active screenreader or just a checkbox may help with this. AFAIK the order for tabbing is controlled by the property "tabIndex".
- Controls that have a tooltip should also get the contents of that tool tip in the property "accessibleDescription".

Making the quite small changes mentioned above will enable blind people to fully use the nitro key. The nitrokey can then be used as alternate solution instead of the well known RSA keys that only have a smal display and are commonly used to do TOTP. So solving this issue may open up new doors for blind people.

**Comment by jans23** [comment]
@szszszsz Since you are reworking the GUI right now, perhaps you could consider this issue and suggested improvements. Thanks.

**Comment by szszszsz** [comment]
Hi!
Qt5 fortunately supports this out-of-the-box. However UI was not designed in the way it could done all work by itself. I had to 'only' rework tab order plus adding proper names and descriptions to changeable controls. I have added read-only labels to tab cycle whenever it could be done in not confusing way to non screen-reader user. On other case names and descriptions were added to fields directly so each one is easily identified. I hope you will be pleased with the results. Feedback is welcome.
All changes tested on Windows 8.1 with mentioned NVDA software.
Feature now staged in beta.3 branch. Issue will be closed when this will be merged to mainline.

**Comment by heikofolkerts** [comment]
Hi,

thanks for the good news. Unfortunately I can only test this when the beta 3 release has been built since I currently haven’t he building infrastructure ready. Could you make a IC build or something else so I can give you feedback ASAP?

 

Heiko

 

 

Von: szszszsz [mailto:notifications@github.com] 
Gesendet: Freitag, 7. April 2017 09:58
An: Nitrokey/nitrokey-app
Cc: heikofolkerts; Author
Betreff: Re: [Nitrokey/nitrokey-app] Nitrokey App should be more accessible to screen reader users (#9)

 

Hi!
Qt5 fortunately supports this out-of-the-box. However UI was not designed in the way it could done all work by itself. I had to 'only' rework tab order plus adding proper names and descriptions to changeable controls. I have added read-only labels to tab cycle whenever it could be done in not confusing way to non screen-reader user. On other case names and descriptions were added to fields directly so each one is easily identified. I hope you will be pleased with the results. Feedback is welcome.
All changes tested on Windows 8.1 with mentioned NVDA software.
Feature now staged in beta.3 branch. Issue will be closed when this will be merged to mainline.

**Comment by szszszsz** [comment]
Sure! We do not have artifacts on our CI build unfortunately.
I will add them to pre-release tag and link here. For which OS would you like the binaries?

**Comment by heikofolkerts** [comment]
Hi,

the binaries please for windows (I am running a windows 10 64bit) but a 32bit shuld be just fine since NVDA is a 32bit application AFAIK.

 

Heiko

 

 

Von: szszszsz [mailto:notifications@github.com] 
Gesendet: Freitag, 14. April 2017 14:59
An: Nitrokey/nitrokey-app
Cc: heikofolkerts; Author
Betreff: Re: [Nitrokey/nitrokey-app] Nitrokey App should be more accessible to screen reader users (#9)

 

Sure! We do not have artifacts on our CI build unfortunately.
I will add them to pre-release tag and link here. For which OS would you like the binaries?

**Comment by szszszsz** [comment]
OK! I will upload test binary in an hour.

**Comment by szszszsz** [comment]
I have uploaded binaries to https://github.com/Nitrokey/nitrokey-app/releases/tag/v1.0-beta.3 .
Let me know how the test went.

**Comment by heikofolkerts** [comment]
Hi,

thanks for the beta version.The app indeed now seems very much more accessible. I found some strange things during testen though:

1.       In the about dialoge all texts are read fine including the serialnumber oft he card. But as far as I remember there was a block containing more detailled information about the card itself. I couldn’t find any oft hat information in the app even when using NVDA object navigation. So Is there more information than one can read with NVDA?

2.       When trying to use the nitrokey to register a TOTP slot for paypal I crashed the nitrokey by 

A)     Generating a random secret in  a new topt slot. Important the input format must be set to base32 at this point.

B)      Opening the paypal form to activate a security key – which is accessible when cancelling the registration of a mobile phone number. That form only accepts a quite short serialnumber – which I expect tob e the „secret“ that syncs both parts.

C)      When copying the secret generated by the nitrokey app tot he paypal form it gets truncated. So I copy nust that part back into the app.

D)     When saving the konfiguration the app crashes I suspect due tot he cut secret. When doing the same using the hex format and removing the last accepted digit so that it results in an secret with a odd number of digits the configuration can be saved. But the swync with paypal isn’t successful with paypal when using two six digits codes. But this may be due tot he limitation from paypal.

 

Heiko

 

 

Von: szszszsz [mailto:notifications@github.com] 
Gesendet: Samstag, 15. April 2017 15:50
An: Nitrokey/nitrokey-app
Cc: heikofolkerts; Author
Betreff: Re: [Nitrokey/nitrokey-app] Nitrokey App should be more accessible to screen reader users (#9)

 

I have uploaded binaries to https://github.com/Nitrokey/nitrokey-app/releases/tag/v1.0-beta.3 .
Let me know how the test went.

**Comment by szszszsz** [comment]
1) Do you mean PIN counters for instance? I will check that. I was sure all labels are accessible with `<Tab>` key.
2a) Could you share the model and firmware version of the device you have used? I would like to retest that. 
2b) Nevertheless, a guard for invalid secret should be added and a warning should be shown.

**Comment by heikofolkerts** [comment]
Hi,

Firmware version 0.7 serialnumber 00003b2f

 

Seems that my mind was playing tricks on me and I remember details about the about dialog that aren’t there. Password counters show up OK.

 

Heiko

 

 

Von: szszszsz [mailto:notifications@github.com] 
Gesendet: Mittwoch, 19. April 2017 18:44
An: Nitrokey/nitrokey-app
Cc: heikofolkerts; Author
Betreff: Re: [Nitrokey/nitrokey-app] Nitrokey App should be more accessible to screen reader users (#9)

 

1.	Do you mean PIN counters for instance? I will check that. I was sure all labels are accessible with <Tab> key.
2a) Could you share the model and firmware version of the device you have used? I would like to retest that.
2b) Nevertheless, a guard for invalid secret should be added and a warning should be shown.

**Comment by szszszsz** [comment]
OK! I guess this could be closed now. Thank you for testing and support!
